### PR TITLE
OSSA: Rewrite address cloning code to fix issues and fix an infinite loop in the ownership verifier.

### PR DESCRIPTION
**Pull request by atrick**
4 commits...

Fix an infinite loop in the OSSA verifier exposed by the new mark_dependence test cases.

Generalize the AccessUseDefChainCloner in MemAccessUtils. It was
always meant to work this way, just needed a client.

Add a new API AccessUseDefChainCloner::canCloneUseDefChain().

Add a bailout for begin_borrow and mark_dependence. Those
projections may appear on an access path, but they can't be
individually cloned without compensating.

Delete InteriorPointerAddressRebaseUseDefChainCloner.

Add a check in OwnershipRAUWHelper for canCloneUseDefChain.

Add test cases for begin_borrow and mark_dependence.

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test source compatibility

**Comment by atrick** [comment]
An unrelated lldb test failed on macos.

**Comment by atrick** [comment]
@swift-ci smoke test macos

**Comment by atrick** [comment]
@swift-ci test macOS

File: include/swift/SIL/MemAccessUtils.h [hunk 1]
```diff
@@ -1497,35 +1499,54 @@ Result AccessUseDefChainVisitor<Impl, Result>::visit(SILValue sourceAddr) {
 
 namespace swift {
 
-/// Clone all projections and casts on the access use-def chain until either the
-/// specified predicate is true or the access base is reached.
+/// Clone all projections and casts on the access use-def chain until the
+/// checkBase predicate returns a valid base.
```
Comment (gottesmm) [inline include/swift/SIL/MemAccessUtils.h:1503]: In a subsequent commit, can you describe the function signature of checkBase here and specify what it means to return a "valid" base. Without that I don't know what that means. I would maybe do something like this:

Clone all projections and cases on the access use-def chain until the checkBase predicate returns a valid SILValue as a base.

File: include/swift/SIL/MemAccessUtils.h [hunk 2]
```diff
@@ -1558,14 +1584,37 @@ class AccessUseDefChainCloner
...
 };
 
-template <typename UnaryPredicate>
+/// Clone all projections and casts on the access use-def chain until the
+/// checkBase predicate returns a valid base.
+///
+/// This will not clone ref_element_addr or ref_tail_addr because those aren't
+/// part of the access chain.
+///
+/// CheckBase is a unary predicate that takes the next source address and either
```
Comment (gottesmm) [inline include/swift/SIL/MemAccessUtils.h:1593]: This is the comment on CheckBase that I wanted above. Maybe add a note above to point here for more information about CheckBase?

File: lib/SIL/Utils/OwnershipUtils.cpp [hunk 3]
```diff
@@ -768,7 +768,7 @@ bool InteriorPointerOperand::findTransitiveUsesForAddress(
       }
 
       // If we are the value use, look through it.
-      llvm::copy(mdi->getValue()->getUses(), std::back_inserter(worklist));
+      llvm::copy(mdi->getUses(), std::back_inserter(worklist));
```
Comment (gottesmm) [inline lib/SIL/Utils/OwnershipUtils.cpp:771]: I am guessing this was the infinite loop? Such a thinko = /.

**Review by gottesmm — APPROVED** [review]
A few comment mixups that I think would make it clearer. Can you do them in a follow on commit?

File: include/swift/SIL/MemAccessUtils.h (hunk 2 above)
Comment (atrick) [reply include/swift/SIL/MemAccessUtils.h:1593]: I had all these comments duplicated, decided the was a bad idea, so added `See comments on the cloneUseDefChain() API.`

File: include/swift/SIL/MemAccessUtils.h (hunk 1 above)
Comment (atrick) [reply include/swift/SIL/MemAccessUtils.h:1503]: The next line below says "See comments on the cloneUseDefChain() API."

**Pull request merged**